src/interface/swarm.cpp

```diff
@@ -472,6 +476,111 @@ void Swarm::Defrag() {
   max_active_index_ = num_active_ - 1;
 }
 
+///
+/// Routine to sort particles by cell. Updates internal swarm variables:
+///  cellSorted_: 1D Per-cell sorted array of swarm memory indices
+///  (SwarmKey::swarm_index_) cellSortedBegin_: Per-cell array of starting indices in
+///  cellSorted_ cellSortedNumber_: Per-cell array of number of particles in each cell
+///
+void Swarm::SortParticlesByCell() {
+  auto pmb = GetBlockPointer();
+  auto swarm_d = GetDeviceContext();
+
+  auto &x = Get<Real>("x").Get();
+  auto &y = Get<Real>("y").Get();
+  auto &z = Get<Real>("z").Get();
+
+  const int nx1 = pmb->cellbounds.ncellsi(IndexDomain::entire);
+  const int nx2 = pmb->cellbounds.ncellsj(IndexDomain::entire);
+  const int nx3 = pmb->cellbounds.ncellsk(IndexDomain::entire);
+
+  auto cellSorted = cellSorted_;
+  auto cellSortedBegin = cellSortedBegin_;
+  auto cellSortedNumber = cellSortedNumber_;
+  int ncells = pmb->cellbounds.GetTotal(IndexDomain::entire);
+  int num_active = num_active_;
+  int max_active_index = max_active_index_;
+
+  // Allocate data if necessary
+  if (cellSortedBegin.GetDim(1) == 0) {
+    cellSortedBegin = ParArrayND<int>("cellSortedBegin_", 0, 0, 0, nx3, nx2, nx1);
+    cellSortedNumber = ParArrayND<int>("cellSortedEnd_", 0, 0, 0, nx3, nx2, nx1);
+  }
+
+  // Write an unsorted list
+  pmb->par_for(
+      "Write unsorted list", 0, max_active_index_, KOKKOS_LAMBDA(const int n) {
+        int i, j, k;
+        swarm_d.Xtoijk(x(n), y(n), z(n), i, j, k);
+        const int cell_idx_1d = i + nx1 * (j + nx2 * k);
+        cellSorted(n) = SwarmKey(cell_idx_1d, n);
+      });
+
+  sort(cellSorted, SwarmKeyComparator(), 0, max_active_index);
+
+  // Update per-cell arrays for easier accessing later
+  const IndexRange &ib = pmb->cellbounds.GetBoundsI(IndexDomain::entire);
+  const IndexRange &jb = pmb->cellbounds.GetBoundsJ(IndexDomain::entire);
+  const IndexRange &kb = pmb->cellbounds.GetBoundsK(IndexDomain::entire);
+  pmb->par_for(
+      "Update per-cell arrays", kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int k, const int j, const int i) {
+        int cell_idx_1d = i + nx1 * (j + nx2 * k);
+        // Find starting index, first by guessing
```

Comment:
This isn't necessarily a very efficient approach, of first guessing where each cell's particles will be in the 1D list, and then marching in O(n) fashion up or down the list until we find the starting particle in that cell, but I couldn't immediately think of a better way to do this and at least it's parallelized on device.

Reply:
What does the literature say about this kind of algorithm? I'd expect that there's previous work on how to solve this in a parallel way.
Also, given that this happens after the sorting step, my naive approach would be to go over all particles in the `par_for` and use an atomic operation to updated the start index for the per cell values.

Reply:
There is a fair amount on this kind of algorithm, but I'm in favor of getting something working now and making it fast later. I think ideally in the long term we would want to rely on `Kokkos::sort`, rather than roll our own solution.

Reply:
oh wait I misunderstood this when just looking at the comments before looking at the code. I deleted my above comment as not relevant.

Reply:
So AMReX uses a counting sort algorithm to organize its particles. I need to read a bit more to understand exactly how that might work here but it might avoid the cost of this kernel, basically.

Reply:
Looking a little more closely I'm not sure the AMReX approach is a clear win relative to what we're doing here, at least looking at https://github.com/AMReX-Codes/amrex/blob/development/Src/Particle/AMReX_DenseBins.H there appears to be a serial region more or less in place of this block of code here.

Similar to the question of atomics in the particles example, is your expectation @pgrete that replacing this with a loop over particles and atomics would be more performant than what I have here even for a large number of particles per cell?

Reply:
As above, only benchmark will (at least somewhat) answer the questions and I expect that the answer heavily depends on whether the particles are sorted (and, as you say, the average number of particles per cell).
If they are sorted, then I'm fairly optimistic that a parallel for region will be faster when written in a way that an individual thread handles N > 1 particles at a time (so that the atomic update of the index for a given individual cell is handles as few threads at possible simultaneously).

Reply:
For this kernel the particles are already sorted (which we assume when guessing the start_index). I actually don't immediately see how to write a kernel that gives each thread multiple particles to work with (right now I'm only looping over all particles or all cells with parthenon kernels) -- are you envisioning some runtime parameter that sets the number of particles per thread in this kernel? That might be a bit esoteric for the user, and unwarranted if this kernel isn't very slow (my guess is it already probably is not a dominant cost)

If this kernel does turn out to be slow, a potentially simpler speedup approach would be to just iteratively guess the start index, by evaluating the expected number of particles the previous guess of `start_index` was off by based on the `cell_idx_1d` of the previously guessed `start_index`

Reply:
> For this kernel the particles are already sorted (which we assume when guessing the start_index). I actually don't immediately see how to write a kernel that gives each thread multiple particles to work with (right now I'm only looping over all particles or all cells with parthenon kernels) -- are you envisioning some runtime parameter that sets the number of particles per thread in this kernel? That might be a bit esoteric for the user, and unwarranted if this kernel isn't very slow (my guess is it already probably is not a dominant cost)

What I had in mind was something like a  `TeamPolicy` where the `N` particles are automatically split across Team and that each thread in a team handles "warp_size" number of indices at a time.
Again, assuming the particles are already sorted should limit the clashes of atomic updates.
I'm happy to table this and leave this discussion for a later when this turns out to be a bottleneck.

My gut feeling is that doing nx1*nx2*nx3 searches (in parallel with potential branch divergence across threads) will be slower than `num_active` parallel atomic updates in many cases.